docs/devcontainerjson-reference.md

```diff
@@ -0,0 +1,172 @@
+
+### Image or Dockerfile specific properties
+
+| Property | Type | Description |
+|----------|------|-------------|
+| `image` | string | **Required** when [using an image](https://code.visualstudio.com/docs/remote/create-dev-container#_using-an-image-or-dockerfile). The name of an image in a container registry ([DockerHub](https://hub.docker.com), [GitHub Container Registry](https://docs.github.com/packages/guides/about-github-container-registry), [Azure Container Registry](https://azure.microsoft.com/services/container-registry/)) that `devcontainer.json` supporting services / tools should use to create the dev container. |
+| `build.dockerfile` / `dockerFile`  | string |**Required** when [using a Dockerfile](https://code.visualstudio.com/docs/remote/create-dev-container#_dockerfile). The location of a [Dockerfile](https://docs.docker.com/engine/reference/builder/) that defines the contents of the container. The path is relative to the `devcontainer.json` file. You can find Dockerfiles for different runtimes in the [vscode-dev-containers repository](https://github.com/microsoft/vscode-dev-containers/tree/main/containers). |
+| `build.context` / `context` | string | Path that the Docker build should be run from relative to `devcontainer.json`. For example, a value of `".."` would allow you to reference content in sibling directories. Defaults to `"."`. |
+| `build.args` | Object | A set of name-value pairs containing [Docker image build arguments](https://docs.docker.com/engine/reference/commandline/build/#set-build-time-variables---build-arg) that should be passed when building a Dockerfile.  Environment and [pre-defined variables](#variables-in-devcontainerjson) may be referenced in the values. Defaults to not set. For example: `"build": { "args": { "MYARG": "MYVALUE", "MYARGFROMENVVAR": "${localEnv:VARIABLE_NAME}" } }` |
+| `build.target` | string | A string that specifies a [Docker image build target](https://docs.docker.com/engine/reference/commandline/build/#specifying-target-build-stage---target) that should be passed when building a Dockerfile. Defaults to not set. For example: `"build": { "target": "development" }` |
+| `build.cacheFrom` | string,<br>array | A string or array of strings that specify one or more images to use as caches when building the image. Cached image identifiers are passed to the `docker build` command with `--cache-from`.
+| `appPort` | integer,<br>string,<br>array |  In most cases, we recommend using the new [forwardPorts property](https://code.visualstudio.com/docs/remote/containers#_always-forwarding-a-port). This property accepts a port or array of ports that should be [published](https://code.visualstudio.com/docs/remote/containers#_publishing-a-port) locally when the container is running. Unlike `forwardPorts`, your application may need to listen on all interfaces (`0.0.0.0`) not just `localhost` for it to be available externally. Defaults to `[]`. |
```

Comment:
This links to a VS Code page. Perhaps we move the text around the difference between publish and forwarding here and link to that instead?

Reply:
Added a section `#### Publishing vs forwarding ports` and linked to that from here. Let me know if the location/linking makes sense to you.